### DISCUSSION TITLE
Intellisense not working for Remote-SSH connection to Arm AArch64 server

### DISCUSSION THREAD

**Discussion by daldr-ntml** [microsoft/vscode-cpptools#12966] [Q&A] [answered]
I am running a simple C++ test program on a remote Arm server using Visual Studio Code Remote-SSH. The server o/s is Ubuntu 24.04 and the compiler is g++-14. The program uses Arm intrinsics defined in:
```
/usr/lib/gcc/aarch64-linux-gnu/14/include/arm_sve.h
```
My .vscode/c_cpp_properties.json contains:
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**",
                "/usr/lib/gcc/aarch64-linux-gnu/14/include/**"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/gcc-14",
            "cStandard": "c17",
            "cppStandard": "c++20",
            "intelliSenseMode": "linux-gcc-arm64",
            "configurationProvider": "ms-vscode.cmake-tools",
            "compilerArgs": [
                "-march=native"
            ]
        }
    ],
    "version": 4
}
```
The CMakeLists.txt is:
```
cmake_minimum_required(VERSION 3.19.4 FATAL_ERROR)    
project(intellisense_test_arm LANGUAGES CXX C)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native")
add_executable(intellisense_test_arm main.cpp)
```
arm_sve.h is included in main.cpp and the build succeeds and runs correctly, but Intellisense is not working for the intrinsics. For example, I see:
![image](https://github.com/user-attachments/assets/2130eb20-f6ff-41c5-b672-1718b2e38eac)

The server has a NVIDIA GH200 Grace Hopper cpu.

How can I fix this please?

**Comment by daldr-ntml** [comment]
Any thoughts on my question please?

**Reply by bobbrow** [reply, marked as answer]
I'm sorry we missed this question.  `arm_sve.h` states the following:
```
/* NOTE: This implementation of arm_sve.h is intentionally short.  It does
   not define the SVE types and intrinsic functions directly in C and C++
   code, but instead uses the following pragma to tell GCC to insert the
   necessary type and function definitions itself.  The net effect is the
   same, and the file is a complete implementation of arm_sve.h.  */
#pragma GCC aarch64 "arm_sve.h"
```

The short answer is that our IntelliSense compiler does not currently process this special `#pragma` and therefore doesn't know to insert these intrinsics into its list of valid types.

You could open a Feature Request to ask us to ask our compiler vendor to add support for these intrinsics.  Since it may be a long while before that request is complete (the vendor is working on c++23 features right now), you could workaround this lack of intrinsics by defining them yourself in a block of code meant only for the IntelliSense compiler to see:

`intrinsics_for_intellisense.h`: (name it whatever you like 😉)
```cpp
#if __INTELLISENSE__
#include <stdint.h>
typedef bool svbool_t;
typedef int8_t svint8_t;
typedef int16_t svint16_t;
typedef int32_t svint32_t;
//  Add any others you need...
#endif
```

Then you'll either need to add this file to your CMake scripts so that it gets included in the IntelliSense configuration (as a forced include), or you need to add the following to your `c_cpp_properties.json`:

```json
{
    "name": "configuration",
    "configurationProvider": "ms-vscode.cmake-tools",
    "forcedInclude": [
        "${workspaceFolder}/intrinsics_for_intellisense.h"
    ],
    "mergeConfigurations": true
}
```
`mergeConfigurations` also adds the `includePath` and `defines` to the configuration, so you'll have to weigh your options to get the configuration you want.

**Reply by daldr-ntml** [reply]
Thanks very much for your answer - very helpful.
